File: python/array/242. Valid Anagram.py
'''
hashmap (own)
Time: O(n))
Space: O(n)
'''

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        hash = {}

        for w in s:
            hash[w] = 1 + hash.get(w, 0)

        for w in t:
            if w not in hash or hash[w] -1 < 0:
                return False 
            hash[w] -= 1

            if hash[w] == 0:
                hash.pop(w)  # del hash[w]

        return True if len(hash) == 0 else False

    def isAnagram(self, s: str, t: str) -> bool:
        return Counter(s) == Counter(t) 

File: javascript/array/242. Valid Anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    const hash = new Map();

    for (let w of s) {
        hash.set(w, (hash.get(w) || 0) + 1);
    }

    for (let w of t){
        if (!hash.has(w) || hash.get(w) === 0) return false;

        hash.set(w, hash.get(w) - 1);
        if (hash.get(w) == 0) hash.delete(w);
    }
    return hash.size <= 0;
};